### ISSUE TITLE
Configure Networking?

### ISSUE THREAD

**Issue by RaymiiOrg** [lenticularis39/axpbox#16]
Do you have any idea how to configure networking? In the video of Stevens Machine Room he uses the prompt on `eth0` and, after installing TCP/IP everything just seems to work, but I can't get that to happen. `eth0` is `enx1062e5f117c9` on my Ubuntu 20.04 machine, and I've compiled with pcap support.

This is in `es40.cfg`:

	pci0.4 = dec21143 
	{
	  adapter = "enx1062e5f117c9" 
	  mac = "08-00-DE-AD-BE-EF"; 
	}

Even if I use an empty config block:

	pci0.4 = dec21143 
	{
	}

and use the prompt question on all different interfaces, no luck with TCP/IP setup. 

I'm sure I've configured that correctly, default gateway, DNS server, correct license, IP address on the interface, hosts database, etc. (`@SYS$STARTUP:TCPIP$CONFIG.COM` and `@SYS$STARTUP:TCPIP$STARTUP`)

Installed pcap libs and recompiled with pcap support:

	$ sudo apt install libpcap-dev 

I even tried to setup a bridge and tap interface, which is what I often do for QEMU/KVM VM's:

As root:

	ip link add br0 type bridge
	ip link set br0 up

	# Replace eth0 with your network interface (ip addr), in my case it was enx1062e5f117c9
	ip link set enx1062e5f117c9 up
	ip link set enx1062e5f117c9 master br0

	# Drop existing IP from eth0, replace 10.0.0.15 by your IP on 
	ip addr del 10.0.0.15/24 dev enx1062e5f117c9
	ip addr flush dev enx1062e5f117c9

	# Assign IP to br0, replace 10.0.0.15 by your IP
	ip addr add 10.0.0.15/24 brd + dev br0
	ip route default via 10.0.0.1 dev br0

	# I had to update my DNS
	systemd-resolve -i br0 --set-dns 10.0.0.101

	# add a tap interface
	ip tuntap add dev tap0 mode tap user remy
	ip link set dev tap0 up
	ip link set tap0 master br0

To resolve this error:

	tap0: You don't have permission to capture on that device (socket: Operation not permitted): /home/remy/tmp/axpbox/src/DEC21143.cpp, line 432
 
I gave the binary correct capabilities:

	sudo setcap cap_net_raw,cap_net_bind_service=+eip ./axpbox

It then stopped complaining, but all further tests I made sure to run as root.

OpenVMS config:

	$ tcpip
	TCPIP> show route
	  
	                             DYNAMIC
	  
	Type           Destination                           Gateway
	  
	AN    0.0.0.0                               10.0.0.1
	AN    10.0.0.0/24                           10.0.0.120
	AH    10.0.0.120                            10.0.0.120
	AH    127.0.0.1                             127.0.0.1

	TCPIP> show interface
	                                                           Packets
	Interface   IP_Addr         Network mask          Receive          Send     MTU
	 
	 LO0        127.0.0.1       255.0.0.0                   0             0    4096
	 WE0        10.0.0.120      255.255.255.0             370           305    1500

Funny thing is, I have a ping running from my linux host, and the data sent received does increase:

	TCPIP> show interface
	                                                           Packets
	Interface   IP_Addr         Network mask          Receive          Send     MTU
	 
	 LO0        127.0.0.1       255.0.0.0                   0             0    4096
	 WE0        10.0.0.120      255.255.255.0             402           327    1500

Pings keep giving a destination unreachable sadly.

On the host I see arp requests: 

	$ sudo tcpdump -nvi br0 dst 10.0.0.120
	tcpdump: listening on br0, link-type EN10MB (Ethernet), capture size 262144 bytes
	19:16:29.950333 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 10.0.0.120 tell 10.0.0.15, length 28
	19:16:30.959119 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 10.0.0.120 tell 10.0.0.15, length 28
	19:16:31.983138 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 10.0.0.120 tell 10.0.0.15, length 28
	19:16:33.007208 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 10.0.0.120 tell 10.0.0.15, length 28

If I force the MAC address on the host:

	arp -i br0 -s 10.0.0.120 08:00:2b:e5:40:00 

No luck:

	$ sudo tcpdump -nvi br0 dst 10.0.0.120 or src 10.0.0.120
	tcpdump: listening on br0, link-type EN10MB (Ethernet), capture size 262144 bytes
	19:19:03.535151 IP (tos 0x0, ttl 64, id 56938, offset 0, flags [DF], proto ICMP (1), length 84)
	    10.0.0.15 > 10.0.0.120: ICMP echo request, id 12, seq 151, length 64
	19:19:04.559115 IP (tos 0x0, ttl 64, id 56974, offset 0, flags [DF], proto ICMP (1), length 84)
	    10.0.0.15 > 10.0.0.120: ICMP echo request, id 12, seq 152, length 64

tcpdump on openvms, after setting the arp mac of the host interface correctly in openvms:

	$ tcpdump -nv 
	tcpdump: Filtering in user process
	tcpdump: listening on WE0, link-type EN10MB (Ethernet), capture size 96 bytes
	# after a minute or so
	 Cancel 

	0 packets captured
	0 packets received by filter
	0 packets dropped by kernel

I do see some code differences in the DEC21143.cpp file from here: https://github.com/JonathanBelanger/es40-ng and here: https://github.com/veprbl/es40 with your one and that of SRM.

**Comment by lenticularis39** [comment]
I also couldn't get network to work, but assumed something's wrong with my setup - however the behavior described by you  makes me think it's definitely a regression. I'll have to look into that.

Thanks for posting the links to other forks of es40 - I somehow missed the fact that Jonathan Belanger made a new fork of es40 (I knew he did it some time ago, but then decided to work on his own emulator instead).

**Comment by RaymiiOrg** [comment]
So far I've tried the following, without luck yet:

- Disable nonblocking on pcap (was a recent change in that corner)
- Replace the DEC21143 related files by the oldest es40 code I could find, except for some formatting nothing's different
- Use the `any` interface for pcap, that segfaults. 

I've got `# define DEBUG_NIC 1` on and I do see some output:

```
21143: nic_write - CSR(7), value: 00000000
21143: nic_write - CSR(7), value: 00000000
21143: nic_write - CSR(0), value: 00000001
%NIC-I-CKSUM: SROM checksum bytes are 5d, 78
21143: nic_write - CSR(0), value: 00009002
21143: nic_write - CSR(13), value: 00000000
21143: nic_write - CSR(6), value: 00000000
21143: nic_write - CSR(5), value: ffffffff
21143: nic_write - CSR(7), value: 00010143
21143: nic_write - CSR(7), value: 00000000
21143: nic_write - CSR(7), value: 00000000
21143: nic_write - CSR(0), value: 00000001
%NIC-I-CKSUM: SROM checksum bytes are 5d, 78

```

Early on in the boot process, in SRM, I see some failure output:

**Internal network loopback Failed**

```
This is serial port #0 on ES40 Emulator


OpenVMS PALcode V1.98-104, Tru64 UNIX PALcode V1.92-105

starting console on CPU 0
initialized idle PCB
initializing semaphores
initializing heap
initial heap 240c0
memory low limit = 1b0000 heap = 240c0, 17fc0
initializing driver structures
initializing idle process PID
initializing file system
initializing hardware
initializing timer data structures
lowering IPL
CPU 0 speed is 800 MHz
create dead_eater
create poll
create timer
create powerup
access NVRAM
Memory size 256 MB
testing memory
..
probe I/O subsystem
probing hose 1, PCI
probing hose 0, PCI
probing PCI-to-ISA bridge, bus 1
bus 0, slot 4 -- ewa -- DE500-BA Network Controller
bus 0, slot 15 -- dqa -- Acer Labs M1543C IDE
bus 0, slot 15 -- dqb -- Acer Labs M1543C IDE
starting drivers
entering idle loop
*** system serial number not set. use set sys_serial_num command.
Partition 0, Memory base: 000000000, size: 010000000
initializing GCT/FRU at 1c8000
Initializing ewa dqa dqb 
Memory Testing and Configuration Status
  Array       Size       Base Address    Intlv Mode
---------  ----------  ----------------  ----------
    0        256Mb     0000000000000000    4-Way

     256 MB of System Memory
Testing the System
Testing the Network
**Internal network loopback Failed**
AlphaServer ES40 Console V7.3-1, built on Feb 27 2007 at 12:57:47
P00>>>boot dqa0

```
Could it be an SRM issue? 

Do you have any pointers where I can look further in the code?

**Comment by RaymiiOrg** [comment]
Here: http://web.archive.org/web/20201031065829/https://astr0baby.wordpress.com/2018/10/11/netbsd-8-on-alpha-es40-simulator/ - the author states that he did get netbsd installed, but failed to get networking working...

**Comment by RaymiiOrg** [comment]
Here is my branch with some tests for reference: https://github.com/RaymiiOrg/axpbox/tree/networking_issue

**Comment by lenticularis39** [comment]
I got exactly the same results when trying network the last time, but now I got further - there is no "Internal network loopback Failed" message in SRM and ARP packets seems to be emitted by the VM when pinging the router (the other two request are unrelated):

```
$ sudo tcpdump -nvi wlp3s0 dst 192.168.1.1 or src 192.168.1.180
dropped privs to pcap
tcpdump: listening on wlp3s0, link-type EN10MB (Ethernet), capture size 262144 bytes
08:00:40.712929 ARP, Ethernet (len 6), IPv4 (len 4), Request who-has 192.168.1.1 tell 192.168.1.180, length 28
08:00:42.778432 ARP, Ethernet (len 6), IPv4 (len 4), Reply 192.168.1.101 is-at a6:d1:46:6c:87:7b, length 28
08:00:43.700010 ARP, Ethernet (len 6), IPv4 (len 4), Reply 192.168.1.101 is-at a6:d1:46:6c:87:7b, length 28
3 packets captured
3 packets received by filter
0 packets dropped by kernel
```

Running with `-e` confirms that the Ethernet addresses of the ARP request are correct:

```
08:07:58.010375 08:00:2b:e5:40:00 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 42: Ethernet (len 6), IPv4 (len 4), Request who-has 192.168.1.1 tell 192.168.1.180, length 28
```

VM configuration (Tru64):

```
# ifconfig -a
lo0: flags=100c88<LOOPBACK,NOARP,MULTICAST,SIMPLEX,NOCHECKSUM>

tu0: flags=c63<UP,BROADCAST,NOTRAILERS,RUNNING,MULTICAST,SIMPLEX>
     inet 192.168.1.180 netmask ffffff00 broadcast 192.168.1.255 ipmtu 1500
```

Maybe attaching to a tap network would work, but still it's a mystery why this works now and didn't before (maybe I tried it back on CentOS, now I use Gentoo).

**Comment by lenticularis39** [comment]
I tried to connect to the libvirt virtual network and ping a VM (in this case the router is the host, and communication between the host and the VM doesn't work properly in some cases when using pcap) and it works:

```
$ sudo tcpdump -nvei virbr0 ether src 08:00:2b:e5:40:00 or ether dst 08:00:2b:e5:40:00
dropped privs to pcap
tcpdump: listening on virbr0, link-type EN10MB (Ethernet), capture size 262144 bytes
08:24:47.868644 08:00:2b:e5:40:00 > ff:ff:ff:ff:ff:ff, ethertype ARP (0x0806), length 42: Ethernet (len 6), IPv4 (len 4), Request who-has 192.168.122.51 tell 192.168.122.180, length 28
08:24:47.868906 52:54:00:3b:a5:1d > 08:00:2b:e5:40:00, ethertype ARP (0x0806), length 42: Ethernet (len 6), IPv4 (len 4), Reply 192.168.122.51 is-at 52:54:00:3b:a5:1d, length 28
08:24:47.898963 08:00:2b:e5:40:00 > 52:54:00:3b:a5:1d, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 9806, offset 0, flags [none], proto ICMP (1), length 84)
    192.168.122.180 > 192.168.122.51: ICMP echo request, id 60929, seq 0, length 64
08:24:47.899514 52:54:00:3b:a5:1d > 08:00:2b:e5:40:00, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 28832, offset 0, flags [none], proto ICMP (1), length 84)
    192.168.122.51 > 192.168.122.180: ICMP echo reply, id 60929, seq 0, length 64
```

```
# ifconfig -a
lo0: flags=100c88<LOOPBACK,NOARP,MULTICAST,SIMPLEX,NOCHECKSUM>

tu0: flags=c63<UP,BROADCAST,NOTRAILERS,RUNNING,MULTICAST,SIMPLEX>
     inet 192.168.122.180 netmask ffffff00 broadcast 192.168.122.255 ipmtu 1500

# ping 192.168.122.51
PING 192.168.122.51 (192.168.122.51): 56 data bytes
64 bytes from 192.168.122.51: icmp_seq=0 ttl=64 time=61 ms
64 bytes from 192.168.122.51: icmp_seq=1 ttl=64 time=30 ms
64 bytes from 192.168.122.51: icmp_seq=2 ttl=64 time=29 ms
^C

----192.168.122.51 PING Statistics----
3 packets transmitted, 3 packets received, 0% packet loss
round-trip (ms)  min/avg/max = 29/40/61 ms
```

**Comment by lenticularis39** [comment]
Traffic that is supposed to be routed out (i.e. into the real local network and/or the Internet) doesn't go through, but I suppose that's because the router is the host.

**Comment by RaymiiOrg** [comment]
In my setup the host (where the es40 vm runs) is not the router. I see you use a wireless interfaces, mine is wired.

I saw in the SMR video that his config has the nic on pci1.1 and I've got it on pci0.4. Changing that to pci1.1 makes no difference for me. One more difference eliminated.

**Comment by lenticularis39** [comment]
I ran AXPbox inside a Debian VM and network works perfectly fine there:

```
# ifconfig tu0 192.168.122.180/24
# ping 192.168.122.1
PING 192.168.122.1 (192.168.122.1): 56 data bytes
64 bytes from 192.168.122.1: icmp_seq=0 ttl=64 time=65 ms
64 bytes from 192.168.122.1: icmp_seq=1 ttl=64 time=25 ms
^C

----192.168.122.1 PING Statistics----
2 packets transmitted, 2 packets received, 0% packet loss
round-trip (ms)  min/avg/max = 25/45/65 ms
# route add default 192.168.122.1
add net default: gateway 192.168.122.1
# ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: icmp_seq=0 ttl=114 time=35 ms
64 bytes from 8.8.8.8: icmp_seq=1 ttl=114 time=39 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=114 time=44 ms
c64 bytes from 8.8.8.8: icmp_seq=3 ttl=114 time=37 ms
^C

----8.8.8.8 PING Statistics----
4 packets transmitted, 4 packets received, 0% packet loss
round-trip (ms)  min/avg/max = 35/39/44 ms
```

This makes me think that maybe it's caused by a firewall, since it is absent in both the Gentoo host and the Debian VM.

**Comment by RaymiiOrg** [comment]
Is the shell output from openvms or from the debian vm?

In my setup I specifically disabled iptables like so: https://raymii.org/s/blog/Reset_iptables_to_ACCEPT_all_-_backup_and_remove_all_existing_rules.html - to exclude that as well.

Do you mean in your debian vm, with axpbox. The networking inside openvms works? Which debian version are you using? Which hypervisor (kvm)? And how is the vm network set up?

**Comment by lenticularis39** [comment]
> Is the shell output from openvms or from the debian vm?
> 

This is from a Tru64 AXPbox VM, sorry for the confusion, should have specified that.

> Do you mean in your debian vm, with axpbox. The networking inside openvms works? Which debian version are you using? Which hypervisor (kvm)? And how is the vm network set up?

I haven't tested OpenVMS yet - I would have to copy the disk image into the VM or set up NFS, and it's quicker to boot up the Tru64 installation CD, where you can drop into a shell and access network. I'm using Debian 10 inside KVM and the VM network is the default network created by libvirt (NAT).

> In my setup I specifically disabled iptables like so: https://raymii.org/s/blog/Reset_iptables_to_ACCEPT_all_-_backup_and_remove_all_existing_rules.html - to exclude that as well.
> 

I tried installing and starting firewalld (that was present on the previous CentOS installation where network didn't work), but it didn't make any difference - 8.8.8.8 (the Google DNS server) is still reachable from the Tru64 VM, so I guess it's not caused by the firewall.

**Comment by lenticularis39** [comment]
I'm using Debian 10.

**Comment by RaymiiOrg** [comment]
Okay so in your debian VM and on your centos host machines, networking inside axpbox works, which adapter do you specify in es40.cfg? Ill see if I can replicate your setup with Tru Unix.

**Comment by lenticularis39** [comment]
> Okay so in your debian VM and on your centos host machines, networking inside axpbox works, which adapter do you specify in es40.cfg? Ill see if I can replicate your setup with Tru Unix. 

AXPbox networking works on my Gentoo host and a Debian VM running on it. Previously the host ran CentOS, where the network didn't work, displaying the same SRM error as in your case.

> Does that require any special setup or is it regular Unix network setup as you've shown above?

It's just regular Unix network setup. No other commands were run except for those shown in the post.

**Comment by lenticularis39** [comment]
I specified ens1:

```
 pci0.4 = dec21143
  {
    adapter = "ens1";
  }
```

**Comment by lenticularis39** [comment]
Pinging the internet working confirmed on OpenVMS in the Debian VM:

```
$ tcpip
TCPIP> ping 8.8.8.8
PING 8.8.8.8 (8.8.8.8): 56 data bytes
64 bytes from 8.8.8.8: icmp_seq=0 ttl=114 time=84 ms
64 bytes from 8.8.8.8: icmp_seq=1 ttl=114 time=48 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=114 time=47 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=114 time=53 ms


----8.8.8.8 PING Statistics----
4 packets transmitted, 4 packets received, 0% packet loss
round-trip (ms)  min/avg/max = 47/58/84 ms
```

**Comment by RaymiiOrg** [comment]
Oh you're the best! I'm going to install a fresh vm with debian 10 and see if I get it working. ens1 is the host adapter in your debian vm? Like eth0?

**Comment by lenticularis39** [comment]
> Oh you're the best! I'm going to install a fresh vm with debian 10 and see if I get it working. ens1 is the host adapter in your debian vm? Like eth0?

It's just a different naming scheme - it has the same meaning as eth0.

> 
> Do you have any clue what changed? I'm beginning to guess something on the host os side,

I have no idea - I might try a standard desktop installation of Fedora/CentOS/Ubuntu in a VM and see if the problem appears there.

**Comment by lenticularis39** [comment]
In any case the advantage of installing in a VM is that you can communicate directly with the (KVM) host (only the communication with the AXPbox host, i.e. the Debian VM, is limited because of pcap limitations) - I was able to telnet into OpenVMS.

**Comment by RaymiiOrg** [comment]
All the kids are asleep, time to install that debian 10 VM. After compiling axpbox and running it, I also no longer see the "network test failed" in the SRM prompt:

```
    0       1024Mb     0000000000000000    4-Way

     1024 MB of System Memory
Testing the System
Testing the Network
AlphaServer ES40 Console V7.3-1, built on Feb 27 2007 at 12:57:47
P00>>>boot dqa0

```

Let's hope this is a good sign.

**Comment by RaymiiOrg** [comment]
One other thing that is different in the debian 10 VM is that I can CTRL+C and axpbox exits:

```
Exiting gracefully: Graceful exit: CTRL-C detected: /root/axpbox/src/System.cpp, line 550
Stop threads: cpu0 srl0 ide0 ide1 nic ali kbd
flash: 2097156 bytes saved.
%FLS-I-SAVEST: Flash state saved to rom/flash.rom
dpr: 16384 bytes saved.
%DPR-I-SAVEST: DPR state saved to rom/dpr.rom
Freeing memory in use by system...
pci0.15(ali_ide).disk0.0(file): Closing file.
pci0.15(ali_ide).disk1.0(file): Closing file.
CPacketQueue(rx_queue): highwater=2, lost=0
root@debian:~/axpbox/build# 
```

The build on my own machine (Ubuntu 20.04) hangs when exiting:

```
^CExiting gracefully: Graceful exit: CTRL-C detected: /home/remy/tmp/axpbox/src/System.cpp, line 550
^C^C^C^C^C^C^C^C^C^C^C^CKilled
```

But it only hangs when I have a fully booted up openvms system that is running a `ping`...

**Comment by lenticularis39** [comment]
> 
> The build on my own machine (Ubuntu 20.04) hangs when exiting:
> 
> ```
> ^CExiting gracefully: Graceful exit: CTRL-C detected: /home/remy/tmp/axpbox/src/System.cpp, line 550
> ^C^C^C^C^C^C^C^C^C^C^C^CKilled
> ```
> 
> But it only hangs when I have a fully booted up openvms system that is running a `ping`...

The current implementation of the function that stops the system waits for all threads to quit. Hanging usually means a thread didn't quit and the main thread is still waiting for it - that could be the Ethernet adapter thread. I'll try to debug this if I'm able to reproduce it.

**Comment by RaymiiOrg** [comment]
Well this is interesting. I configured the debian 10 virtualbox VM first with 1 NIC, bridged to my own nic. That gave the same result, no connection from either my host or debian 10 vm to openvms. 

Then I did as you suggested, added an extra NIC, virtualbox NAT, and configured that in es40.cfg. Then in OpenVMS, set the correct network config (address/gateway):

```
$ tcpip show route
  
                             DYNAMIC
  
Type           Destination                           Gateway
  
AN    0.0.0.0                               10.0.2.2
AN    10.0.2.0/24                           10.0.2.20
AH    10.0.2.20                             10.0.2.20
AH    127.0.0.1                             127.0.0.1
$ tcpip show interface
                                                           Packets
Interface   IP_Addr         Network mask          Receive          Send     MTU
 
 LO0        127.0.0.1       255.0.0.0                   0             0    4096
 WE0        10.0.2.20       255.255.255.0             167           170    1500
$ 
```

After which it works, internet connection and local (vbox nat) ping and DNS:

```
$ tcpip ping raymii.org
PING raymii.org (199.19.225.3): 56 data bytes
64 bytes from 199.19.225.3: icmp_seq=0 ttl=63 time=187 ms
64 bytes from 199.19.225.3: icmp_seq=1 ttl=63 time=157 ms
64 bytes from 199.19.225.3: icmp_seq=2 ttl=63 time=162 ms
64 bytes from 199.19.225.3: icmp_seq=3 ttl=63 time=169 ms


----raymii.org PING Statistics----
4 packets transmitted, 4 packets received, 0% packet loss
round-trip (ms)  min/avg/max = 157/169/187 ms
$ 
$ tcpip ping 10.0.2.2
PING 10.0.2.2 (10.0.2.2): 56 data bytes
64 bytes from 10.0.2.2: icmp_seq=0 ttl=64 time=25 ms
64 bytes from 10.0.2.2: icmp_seq=1 ttl=64 time=18 ms
64 bytes from 10.0.2.2: icmp_seq=2 ttl=64 time=12 ms
64 bytes from 10.0.2.2: icmp_seq=3 ttl=64 time=16 ms


----10.0.2.2 PING Statistics----
4 packets transmitted, 4 packets received, 0% packet loss
round-trip (ms)  min/avg/max = 12/18/25 ms
$ 
```

I can however, **not** ping the debian VM on which axpbox is running, and neither can the debian VM ping openvms:

openvms to debian:

```
$ tcpip ping 10.0.2.15
PING 10.0.2.15 (10.0.2.15): 56 data bytes


----10.0.2.15 PING Statistics----
4 packets transmitted, 0 packets received, 100% packet loss
%SYSTEM-F-TIMEOUT, device timeout
$ 
```

Via virtualbox I set up a port forward, on the debian VM to the OpenVMS IP, then I can telnet from my ubuntu host (on which debian vm runs) to OpenVMS:

Port forward:

![Selectie_2020 10 31 07](https://user-images.githubusercontent.com/1927494/97780026-51fa7900-1b82-11eb-8db9-c2f481de51d1.png)

Ubuntu host: 

```

$ telnet 127.0.0.1 2323
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.


 Welcome to OpenVMS (TM) Alpha Operating System, Version V8.4    

Username: SYSTEM
Password: 
   Welcome to OpenVMS (TM) Alpha Operating System, Version V8.4
    Last interactive login on Wednesday, 17-OCT-2018 22:02:44.02
$ 
```

OpenSSH as well after configuring it:

```
[14:08:51] [remy@gateway] [ ~/tmp/axpbox/build (veprbl-and-mvorl-patches) ]
$ ssh -o HostKeyAlgorithms=ssh-rsa,ssh-dss -o KexAlgorithms=diffie-hellman-group1-sha1 -o Ciphers=aes128-cbc,3des-cbc -o MACs=hmac-md5,hmac-sha1 -l SYSTEM -p 2222 127.0.0.1
SYSTEM@127.0.0.1's password: 

    Last interactive login on Wednesday, 17-OCT-2018 22:13:01.77
    Last non-interactive login on Wednesday, 17-NOV-1858 00:00:00.00
$ show sys
OpenVMS V8.4  on node U1   17-OCT-2018 22:17:26.90   Uptime  0 00:16:37
  Pid    Process Name    State  Pri      I/O       CPU       Page flts  Pages
00000101 SWAPPER         HIB     16        0   0 00:00:00.14         0      4   
00000104 LANACP          HIB     14       78   0 00:00:00.13       118    147   
00000106 FASTPATH_SERVER HIB     10        9   0 00:00:00.07        78     95   
00000107 IPCACP          HIB     10        9   0 00:00:00.02        37     51   
00000108 ERRFMT          HIB      8       56   0 00:00:00.13       121    130   
0000010A OPCOM           HIB      7       86   0 00:00:00.12       100     50   
0000010B AUDIT_SERVER    HIB     10       61   0 00:00:00.13       140    148   
0000010C JOB_CONTROL     HIB      9      148   0 00:00:00.10        50     76   
0000010F SECURITY_SERVER HIB     10       66   0 00:00:00.41       375    376   
00000111 ACME_SERVER     HIB     10       71   0 00:00:00.73       449    449 M 
0000011D SMHANDLER       HIB      8       52   0 00:00:00.20       177    156   
0000011E SYSTEM          LEF      4    17529   0 00:00:44.53     17513     90   
00000139 TCPIP$INETACP   HIB      8      184   0 00:00:00.53       369    303   
00000144 _TNA2:          LEF      4       62   0 00:00:00.33       146    111   
00000148 TCPIP$SSH_BG97  LEF      9      654   0 00:00:04.00       790    718  N
00000149 _FTA1:          CUR   0  4       62   0 00:00:00.32       187    158   
$ 
```

**Comment by RaymiiOrg** [comment]
I don't quite understand what is going wrong here, or, why the setup with the virtualbox NAT network works, besides the communication between the debian vm axpbox runs on and openvms inside axpbox.

I'd like it to work without a port forward inside virtualbox, but nontheless, I have network and can now go test with clustering.

I'll try to write a guide in the next few days documenting this setup to at least get some networking working (hehe).

Not marking the issue as solved, but still, thank you very much for your help so far!

**Comment by lenticularis39** [comment]
> 
> Then I did as you suggested, added an extra NIC, virtualbox NAT, and configured that in es40.cfg. Then in OpenVMS, set the correct network config (address/gateway):
> 

Maybe there's a subtle detail that prevents the traffic from going through on physical/bridged interfaces. I tried it on my PowerBook G4, connected to the router via its Sun GEM Ethernet port, but the CPU is big endian, which caused some issues - I might try it on my Raspberry Pi instead.

> 
> I can however, **not** ping the debian VM on which axpbox is running, and neither can the debian VM ping openvms:
> 

I think that's a limitation of pcap-based networking - I remember reading something like that in the documentation of SimH.

**Comment by lenticularis39** [comment]
> I don't quite understand what is going wrong here, or, why the setup with the virtualbox NAT network works, besides the communication between the debian vm axpbox runs on and openvms inside axpbox.
> 
> I'd like it to work without a port forward inside virtualbox, but nontheless, I have network and can now go test with clustering.
> 
> I'll try to write a guide in the next few days documenting this setup to at least get some networking working (hehe).
> 
> Not marking the issue as solved, but still, thank you very much for your help so far!

Thank you for your help.

**Comment by RaymiiOrg** [comment]
You're right about pcap and the simh docs:

> On Windows using the WinPcap interface, the simulated computer can "talk" to
the host computer on the same interface.  On other platforms with libpcap 
(*nix), the simulated computer can not "talk" to the host computer via the 
selected interface, since simulator transmitted packets are not received
by the host's network stack. The workaround for this is to use a second NIC 
in the host and connect them both into the same network; then the host and 
the simulator can communicate over the physical LAN.

I'll go and test what happens if I add another nic to my desktop  (or the VM). If the nat part is not required that simplifies the setup and no port forwarding is required.

**Comment by RaymiiOrg** [comment]
Tested the other NIC on my desktop (USB adapter), when I bridge to that in Virtualbox it works (from my desktop to debian VM -> OpenVMS) 

When I run AXPbox on my desktop (no debian VM) and configure that USB NIC inside `es40.cfg`, it works fine as well.

When I have 2 adapters in virtualbox bridged to my regular nic, it doesn't work, neither inside OpenVMS nor desktop->openVMS. 

So it's either 2 actual seperate network interfaces, or the Virtualbox NAT for me.

**Comment by RaymiiOrg** [comment]
I think the issue can be resolved, I'll go and make a document out of it for the Wiki. If you want to leave it open for more testing/debugging, that's fine by me as well, I can help test/develop there.

**Comment by lenticularis39** [comment]
OK, it's clear now that there wasn't any regression. Closing the issue for now - if a problem is encountered in the future it can be re-opened.

**Comment by RaymiiOrg** [comment]
First version of my attempt at documenting this is here: https://github.com/lenticularis39/axpbox/wiki/OpenVMS-Networking-setup

**Comment by doktor-why** [comment]
I was able to setup it up, including the network. I am using a Centos8 system with axpbox running VMS. I was a bit confused with the network setup. I used a bridged NIC adapter in my Centos8 system "bridge0" and also the single real adapter underneath the bridge0, "eno1". I also used another NIC "eno3". No issues with the network. I can ping and telnet (somewhat). 

My issue is that the virtual NIC seems to have a problem with packet size. If I telnet, I can login fine, I can issue dir command and that's when problems start.  Telnet hangs after few screens from the dir command. FTP hangs a lot too.

What I see is this in the console output after the ftp hang or telnet is the following. it seems that it's getting oversize packets, maybe?

ide1.0 aborting on command 0xec
*** IP interrupt set for CPU 0 from CPU 0(@ 110d5)
*** IP interrupt cleared for CPU 0 from CPU 0(@ 1a140c).
ide1.0 aborting on command 0xec
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc
CPacketQueue(rx_queue):add() packet lost! Size = 2974.. dst: 08-00-2b-e5-40-00 .. src: e0-63-da-23-ae-dc


I wonder if anyone has such issue? and more importantly how make it go away.
